### PR TITLE
add proposal that updates core design w.r.t volume controller, & policy

### PR DESCRIPTION
**Pull request by AmitKumarDas**
This is a design proposal w.r.t:

- volume controller implementation,
- storage policy implementation,
- declarative code,
- infrastructure as a code

### CODE REVIEW

**Comment by kmova** [comment]
The requirement needs to factor in -- supportability or compatbility of different versions of k8s, how should the volume - k8s be specified. Also a note on the implementation, if storage policies are implented via CRDs, should they be also represented via ConfigMaps -- I see ConfigMaps -- as one form of CRDs.

**Comment by AmitKumarDas** [comment]
@kmova Supporting multiple versions of K8s is dependant (delegated to) client-go.
We need to be active in client-go that supports multiple versions of K8s perhaps 5 versions.

maya can then use as many versions of K8s that client-go supports.

Note: In this design, the volume is split into properties & policies & these are in turn split into granular properties & policies. 

- These finely granular policy are mapped 1:1 to K8s Kind or a fragment of K8s Kind.
- Similarly, finely granular property are mapped 1:1 to cstor/jiva/etc config

Different K8s versions across clusters/customers and requirement of different versions of a policy in a single K8s deployment are something we need to care about.

Lets refer our policy & property config maps as resources.

I feel we should never ever delete/update our resources. Rather we create the modifications in a new label which is in turn a create resource. Since volume is nothing but name, capacity & reference to these resources, volume should be able to support backward compatibility & what not due to the assumptions we are making for our resources i.e. never delete or update.

File: openebs-features/VOLUME_3.0.md [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: cm-storagepolicy-abc:0.5.0
+data:
+  category: openebs.io/storage-policy
+  type: StoragePolicy
+  version: 0.5.0
```
Comment (AmitKumarDas) [inline openebs-features/VOLUME_3.0.md]: rename **version** to **typeversion**

File: openebs-features/VOLUME_3.0.md [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: cm-storageproperty-default:0.5.0
+data:
+  category: openebs.io/storage-property
+  type: StorageProperty
+  version: 0.5.0
```
Comment (AmitKumarDas) [inline openebs-features/VOLUME_3.0.md]: rename **version** to **typeversion**

File: openebs-features/VOLUME_3.0.md [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: cm-config-cstor-default:0.5.0
+data:
+  category: property.openebs.io/config
+  type: CStorConfig
+  version: 0.5.0
```
Comment (AmitKumarDas) [inline openebs-features/VOLUME_3.0.md]: rename **version** to **typeversion**

File: openebs-features/VOLUME_3.0.md [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+}
+
+// Properties holds various references using which volume's properties
+// can be determined
+type Properties struct {
+  // ConfigMap is the name of the ConfigMap that holds volume's properties
+  ConfigMap   string
+  
+  // EndPoint is the url of the service that provides volume's properties
+  EndPoint    string
```
Comment (AmitKumarDas) [inline openebs-features/VOLUME_3.0.md]: On second thoughts, I feel EndPoint should be eliminated.
It should not be a part of maya volume controller logic.
Rather, endpoints should be pulled to get the entire set of policies & get these policies installed on the env.

NOTE: IMHO Pulling these policies can be manual or some controller other than volume controller.
NOTE: Pulling of these policies can be a job of Policy CRD controller
NOTE: Operator should have set some CronJob to build a storageclass by referring to POLICY CRD 
  - Policy CRD controller will get the latest recommend policy from an endpoint & install it
  - Policy CRD controller will then create a storageclass that refers to these new policies & install it
NOTE: EndPoints should have done due verification & provide some performance numbers of this policy that is being pulled as a latest recommendation.
  - If Policy CRD controller finds the performance numbers to be BAD than the current policy it has then it need not install these policies & storageclass

There has to be some common ground/logic w.r.t Policy CRD & Endpoint to publish policy ranks/performance.

File: openebs-features/VOLUME_3.0.md [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: cm-ctrl-svc-jiva:0.5.0
+data:
+  category: policy.openebs.io/controller-service
+  type: K8sService
+  version: 0.5.0
```
Comment (AmitKumarDas) [inline openebs-features/VOLUME_3.0.md]: rename **version** to **typeversion**

File: openebs-features/VOLUME_3.0.md [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: cm-ctrl-plc-jiva:0.5.0
+data:
+  category: policy.openebs.io/controller-placement
+  type: K8sDeployment
+  version: 0.5.0
```
Comment (AmitKumarDas) [inline openebs-features/VOLUME_3.0.md]: rename **version** to **typeversion**

File: openebs-features/VOLUME_3.0.md [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: cm-rep-plc-jiva:0.5.0
+data:
+  category: policy.openebs.io/replica-placement
+  type: K8sDeployment
+  version: 0.5.0
```
Comment (AmitKumarDas) [inline openebs-features/VOLUME_3.0.md]: rename **version** to **typeversion**

**Comment by AmitKumarDas** [comment]
@kmova 
>Also a note on the implementation, if storage policies are implented via CRDs, should they be also represented via ConfigMaps -- I see ConfigMaps -- as one form of CRDs.

It need not be. The main StoragePolicy CRD should have ability to refer to granular policies.
The main StoragePolicy CRD can have some defined properties in addition to above in-direct references.
The defined & direct properties can be some performance nos (avg, best, worst), no of PVs it is associated with, no of PVs that fall in avg category, etc. I guess a full blown query system might be required. Not sure if this kind of query should be a part of Policy CRD schema.

**Comment by AmitKumarDas** [comment]
https://docs.google.com/document/d/1cLPGweVEYrVqQvBLJg6sxV-TrE5Rm2MNOBA_cxZP2WU/edit#

>Push: facilitates separation of concerns and late binding
can be explicit, such as with kubectl set or HorizontalPodAutoscaler
can be implicit, such as with LimitRange, PodSecurityPolicy, PodPreset, initializers
good for attaching policies to selected resources within a scope (namespace and/or label selector)

This talks about applying policies

**Comment by AmitKumarDas** [comment]
https://kubernetes.io/docs/concepts/api-extension/custom-resources/

Above links describes when to use a CR, when to use declarative API vs. RPC, when to use ConfigMap vs CR !!!

**Comment by AmitKumarDas** [comment]
https://github.com/istio/istio/blob/master/mixer/doc/adapters.md

This one is the policy implementer w.r.t networking !!

**Comment by AmitKumarDas** [comment]
This has been done as part of volume policy.

**Pull request closed without merging**